text/0240-app-telemetry-schema-vision-roadmap.md

```diff
@@ -0,0 +1,305 @@
+
+The internal catalog is used to define all the attributes and metrics in this
+artifact. This design allows for the reuse of the same attributes or metrics
+multiple times in different signals. It is expected to be a very common pattern
+to reuse the same subset of attributes or metrics across several signals.
+
+A dedicated OTEP will precisely define the structure and format of the Resolved
+Telemetry Schema.
+
+## Roadmap
+
+To facilitate the review process and progressively deliver value to the
+OpenTelemetry community, a series of OTEPs and tools are proposed.
+
+### OTEPs
+
+* Telemetry Schema versions (incremental steps):
+  * V1.1.1: Add a section to import one or several Semantic Convention
+  Registries. Once implemented, the official Semantic Convention Registry will
+  be referenced from the official OpenTelemetry Root Telemetry Schema.
+  * V1.2: Add references to parent schemas and define a schema section to
+  describe application/library signals. In this version, only metrics, logs, and
+  spans signals are described. This is the first version of the telemetry schema
+  that can be used by the Client SDK generator. In this version only one schema
+  can be inherited. More complex hierarchies will be supported in version 1.5.
+  * V1.3: Add support for events in the schema section. This OTEP is contingent
+  upon the approval of events in OpenTelemetry.
+  * V1.4: Add support for metric_groups (multivariate metrics) in the schema
+  section. This will only be relevant if a support for multivariate metrics is
+  planned in the Client SDK Generator.
+  * V1.5: Add support for multi-inheritance and for schema hierarchies of more
+  than 2 levels. This is useful for supporting complex environments and will
+  allow the distribution of semantic convention registries across vendors and
+  enterprises. This OTEP depends on the “Schema Child/Parent relationship
+  resolution” OTEP.
+* Resolved Schema Format v1.0.
+* Schema-driven Client SDK Generator.
+* Schema Child/Parent relationship resolution.
+
+### Tools
+
+OpenTelemetry tools
+
+* Semantic Convention Registry Merger (Johannes Tax's proposal): This tool will
+merge all the semantic convention files of a registry into a single and
+referencable file.
+* Search Engine: This tool will provide search capability within semantic
+convention registries and telemetry schemas.
+  * V1 focusing on the semantic convention registry.
+  * V2 focusing on telemetry schema v1.2.
+* Telemetry Schema Resolver: This tool will resolve the telemetry schema and its
+dependencies. The main outcome of this tool is a Resolved Telemetry Schema.
+* Schema-driven Client SDK Generator: This tool will generate a client SDK or a
+client API from a telemetry schema v1.2. It will be designed to be extensible,
+allowing each language SIG and the general community to create templates for
+generating client SDKs and APIs in specific programming languages.
```

Comment:
This is conceptually similar to the java [instrumentation-api](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation-api). The docs on instrumentation-api are sparse, but the general idea is to provide a higher order API built on top of the metric, trace, log API (only metric and trace in practice) which makes it easy to build instrumentation which conforms to the semantic conventions. As an instrumentation author, you implement a hand full of interfaces to extract attributes relevant to the semantic convention being implemented, and the instrumentation API takes care of recording spans / metrics correctly.

Reply:
I may not be familiar enough with the Java Instrumentation API, but my understanding so far was that the Instrumentation API contains all the attributes and metrics defined in the official semantic convention registry. So, in a way, it's not code generated for a specific application context (or library) but rather code generated for a registry or a catalog of conventions. If my interpretation is correct, then I think that this OTEP follows a somewhat different approach, although complementary.